source/crud/tests/README.md

```diff
@@ -201,3 +201,357 @@ CommandSucceededEvents. Then, insert an invalid document (e.g. `{x: 1}`) and ass
 code is `121` (i.e. DocumentValidationFailure), and that its `details` property is accessible. Additionally, assert that
 a CommandSucceededEvent was observed and that the `writeErrors[0].errInfo` field in the response document matches the
 WriteError's `details` property.
+
+### 3. `MongoClient.bulkWrite` batch splits a `writeModels` input with greater than `maxWriteBatchSize` operations
+
+Test that `MongoClient.bulkWrite` properly handles `writeModels` inputs containing a number of writes greater than
+`maxWriteBatchSize`.
+
+This test must only be run on 8.0+ servers.
+
+Construct a `MongoClient` (referred to as `client`) with
+[command monitoring](../../command-logging-and-monitoring/command-logging-and-monitoring.rst) enabled to observe
+CommandStartedEvents. Perform a `hello` command using `client` and record the `maxWriteBatchSize` value contained in the
+response. Then, construct the following write model (referred to as `model`):
+
+```json
+InsertOne: {
+  "namespace": "db.coll",
+  "document": { "a": "b" }
+}
+```
+
+Construct a list of write models (referred to as `models`) with `model` repeated `maxWriteBatchSize + 1` times.
+Execute `bulkWrite` on `client` with `models`. Assert that the bulk write succeeds and returns a `BulkWriteResult`
+with an `insertedCount` value of `maxWriteBatchSize + 1`.
+
+Assert that two CommandStartedEvents (referred to as `firstEvent` and `secondEvent`) were observed for the `bulkWrite` command.
+Assert that the length of `firstEvent.command.ops` is `maxWriteBatchSize`. Assert that the length of `secondEvent.command.ops`
+is 1. If the driver exposes `operationId`s in its CommandStartedEvents, assert that `firstEvent.operationId` is equal to
+`secondEvent.operationId`.
+
+### 4. `MongoClient.bulkWrite` batch splits when an `ops` payload exceeds `maxMessageSizeBytes`
```

Comment:
Consider adding a test for the following scenario: a write with an as-yet-unseen namespace is encoded into ops, and maxMessageSizeBytes is not yet exceeded, but the addition of one more element to nsInfo causes maxMessageSizeBytes to be exceeded. In that case, the driver has to remove the last element from both ops and nsinfo.

Reply:
It may be brittle, but the consequence of getting this wrong could be sending a message to the server that is too large and will be rejected.  Without a test for it, I fear that we will have a bunch of difficult-to-find bugs in drivers around this edge case.

Reply:
Was able to make this work with some less-brittle numbers with the 1000 byte overhead approximation.